Fix #10155 include writer outputstream

gregw opened the pull request:
Handle including when there is a mix of input and output streams

gregw committed:
Including a static resource #10142

Tests to replicate the problem in #10142

lorban committed:
#10142 - tentative fix for included static resources

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10142 - tentative 2

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12.0.x/10142-include-static

lorban committed:
#10142 - tentative 3

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Fix #10155 Include mixed writers and input streams

Fix #10155 Include mixed writers and input streams

gregw commented:
This also includes the commits from #10154

joakime commented:
This PR has 2 testcase failures.

* org.eclipse.jetty.ee10.jstl.JspIncludeTest.testTopWithIncluded
* org.eclipse.jetty.ee10.servlet.IncludedServletTest.testTopWithIncludedHeader

gregw committed:
Fix #10155 Include mixed writers and input streams

If the included servlet uses the writer, we must flush it and thus commit the response.  Previously we were sometimes able to avoid the commit, but that is too complex now and not required by the spec.

gregw committed:
Fix #10155 Include mixed writers and input streams

Avoided extra buffer

gregw committed:
Merge branch 'jetty-12.0.x' into fix/12.0.x/10155-include-writer-outputstream

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/IncludedServletTest.java:
```diff
@@ -172,6 +172,9 @@ public void testTopWithIncludedHeader() throws IOException
             IO.copy(reader, writer);
 
             String response = writer.toString();
+            connection.getHeaderFields().entrySet().forEach(System.err::println);
```
These `System.err.println` should go away.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DispatcherTest.java:
```diff
@@ -1078,22 +1239,50 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+
+    public static class HelloServlet extends HttpServlet implements Servlet
+    {
+        // The logic linked to this field be deleted and the writer always used once #10155 is fixed.
```
If you want to keep the `useWriter` logic, this comment should go away.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DispatcherTest.java:
```diff
@@ -380,10 +384,165 @@ public void testInclude() throws Exception
+            """);
+
+        String expected = """
+            HTTP/1.1 200 OK\r
```
What's the reason why using the writer results in the absence of the `Content-Length` header?

gregw replied:
The setting of a content-length header is highly dependent on the HTTP version, the request version, the request headers and most importantly when/how the response is committed.    We don't want to test if HttpConnection and HttpGenerator are doing the right framing here, that is tested elsewhere.  We just want to see content and/or headers

gregw committed:
Fix #10155 Include mixed writers and input streams

Avoid flushing writer if at all possible to avoid early commit
